aten/src/ATen/native/cuda/Loss.cu

```diff
@@ -153,4 +154,260 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+    int64_t* target,
+    scalar_t* output,
+    scalar_t* weights,
+    int n_classes,
+    int ignore_index) {
+  CUDA_KERNEL_LOOP(index, batch_size) {
+    int cur_target = target[index];
+    if (cur_target == ignore_index) {
+      output[index] = static_cast<scalar_t>(0);
+      continue;
+    }
+    CUDA_KERNEL_ASSERT(cur_target >= 0 && cur_target < n_classes);
+    auto cur_weight =
+        weights != nullptr ? weights[cur_target] : static_cast<scalar_t>(1);
+    output[index] = -cur_weight * input[index][cur_target];
+  }
+}
+
+template <typename scalar_t>
+__global__ void nll_loss_forward_reduce_cuda_kernel_1d(
+    scalar_t* output,
+    scalar_t* total_weight,
+    scalar_t* input,
+    int64_t* target,
+    scalar_t* weights,
+    bool size_average,
+    int n_classes,
+    int64_t ignore_index) {
+  CUDA_KERNEL_ASSERT(threadIdx.x == 0 && threadIdx.y == 0 & threadIdx.z == 0);
+
+  int t = static_cast<int>(*target);
+  if (t != static_cast<int>(ignore_index)) {
+    CUDA_KERNEL_ASSERT(t >= 0 && t < n_classes);
+    scalar_t cur_weight =
+        weights != nullptr ? weights[t] : static_cast<scalar_t>(1);
+    *output = -cur_weight * input[t];
+    *total_weight = cur_weight;
+    if (size_average && *total_weight > 0) {
+      *output /= *total_weight;
+    }
+  }
+}
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void nll_loss_forward_reduce_cuda_kernel_2d(
+    scalar_t* output,
+    scalar_t* total_weight,
+    scalar_t* input,
+    int64_t* target,
+    scalar_t* weights,
+    bool size_average,
+    int nframe,
+    int ndim,
+    int n_classes,
+    int64_t ignore_index) {
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  __shared__ accscalar_t sh_inputs[NLL_LOSS_THREADS],
+      acc_weight[NLL_LOSS_THREADS];
+
+  sh_inputs[threadIdx.x] = static_cast<accscalar_t>(0);
+  acc_weight[threadIdx.x] = static_cast<accscalar_t>(0);
+  for (int i = threadIdx.x; i < nframe; i += NLL_LOSS_THREADS) {
+    int t = target[i];
+    if (t != static_cast<int>(ignore_index)) {
+      CUDA_KERNEL_ASSERT(t >= 0 && t < n_classes);
+      scalar_t cur_weight =
+          weights != nullptr ? weights[t] : static_cast<scalar_t>(1);
+      sh_inputs[threadIdx.x] -= input[i * ndim + t] * cur_weight;
+      acc_weight[threadIdx.x] += cur_weight;
+    }
+  }
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    *output = *total_weight = static_cast<scalar_t>(0);
+    accscalar_t output_acc = 0;
+    accscalar_t total_weight_acc = 0;
+    for (int i = 0; i < NLL_LOSS_THREADS; ++i) {
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
+    if (size_average) {
+      if (nframe == 0) {
+        // Mean reduction on empty tensors produces NaN
+        *output = std::numeric_limits<double>::quiet_NaN();
+      }
+      if (total_weight_acc != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
```

Comment:
you need `else` here, otherwise `output` won't be set for total_weight_acc = 0, and please add tests for all these cases.

Reply:
The original implementation had `*output` initialized to 0, which happened to be correct when `total_weight_acc==0`.

I redid the implementation a little to only set `*output` once and added test for the `nan` case and the `size_average && total_weight_acc == 0` case.